doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,97 @@
+How to analyze Haskell performance
+==================================
+
+When a Haskell application is slow or uses too much memory,
+Cabal and the
+`Haskell GHC compiler <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
+can help you understand why.
+
+The main steps are to
+1. let GHC insert performance measuring code into your application,
+2. run the application to collect a performance report and
+3. visualize that report.
+
+The process of inserting performance measuring code and collecting performance information
+is called "profiling" and Cabal makes this is easy to do:
+
+Profiling CPU performance
+-------------------------
+
+First, build your application, e.g. ``my-app``, with profiling enabled:
+
+.. code-block:: console
+
+    $ cabal build --enable-profiling --profiling-detail=late exe:my-app
+
+Under the hood this will pass the corresponding
+`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
+to GHC. With the additional Cabal flag ``--profiling-detail=late`` GHC is instructed to use
+`"late-cost-center" profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
+and insert performance measuring code only after important optimisations
+have been applied to minimize the impact of profiling on performance.
+See the Cabal section on :ref:`profiling options <profiling-options>` for more details.
+
+Second, run the application with the necessary
+`runtime system (RTS) options <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
+to produce a profile report.
+
+.. code-block:: console
```

Comment:
Do we want to suggest `$(cabal list-bin my-app) +RTS -pj -RTS` instead of needing to copy the path? P personally use that idiom fairly often when testing changes.

Reply:
Good idea, added.